In [1]:
import pandas as pd
import numpy  as np
import warnings

warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv('datesets/uk_road_accident.csv')

In [5]:
df


Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [7]:
df.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h2>Clean the dataset</h2>

In [18]:
df['Latitude'] = df['Latitude'].fillna(df['Latitude'].mean())
df['Longitude'] = df['Longitude'].fillna(df['Longitude'].mean())
df['Road_Surface_Conditions'] = df['Road_Surface_Conditions'].fillna(df['Road_Surface_Conditions'].mode()[0])
df['Road_Type'] = df['Road_Type'].fillna(df['Road_Type'].mode()[0])
df['Urban_or_Rural_Area'] = df['Urban_or_Rural_Area'].fillna(df['Urban_or_Rural_Area'].mode()[0])
df['Weather_Conditions'] = df['Weather_Conditions'].fillna(df['Weather_Conditions'].mode()[0])

In [19]:
df.isnull().sum()


Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h2>1. Which accident severity occurs most often?</h2>

In [20]:
df['Accident_Severity'].value_counts()


Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<H2>Insight: "Slight" accidents are the most common, followed by "Serious" and "Fatal".</H2>

<h2>2. Which district reports the most accidents?</h2>

In [50]:
df[df['Accident_Severity'] == 'Fatal']['District Area'].value_counts()

District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
                           ... 
Havant                        2
Orkney Islands                2
Harlow                        1
Clackmannanshire              1
Stevenage                     1
Name: count, Length: 422, dtype: int64

<h2>Insight: The busiest district has the highest accident count, showing a hotspot of risk.</h2>

<h2>3. What is the most common light condition in accidents?</h2>

In [49]:
df['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>Insight: Most accidents happen in daylight.</h2>

<h2>4. What is the most common weather condition?</h2>

In [48]:
df['Weather_Conditions'].value_counts()


Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>Insight: Most accidents happen in fine weather with no high winds.</h2>

<h2>5. Which road surface condition is most frequent?</h2>

In [47]:
df['Road_Surface_Conditions'].value_counts()


Road_Surface_Conditions
Dry                     448547
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

<h2>Insight: Majority of accidents occur on dry roads.</h2>

<h2>6. What is the average number of vehicles per accident?</h2>

In [46]:
df['Road_Type'].value_counts()


Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h2>Insight: Single carriageways are the most common road type for accidents.</h2>

<h2>7. What is the average number of casualties per accident?</h2>

In [17]:
df['Urban_or_Rural_Area'].value_counts()


Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insight: Urban areas report far more accidents than rural areas.</h2>

<h2>8. Which district has the highest total casualties?</h2>

In [19]:
df['Vehicle_Type'].value_counts()


Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>Insight: Cars are by far the most common vehicles in accidents.</h2>

<h2>9. Which district has the highest average casualties per accident?</h2>

In [24]:
df.groupby('District Area')['Number_of_Casualties'].mean().sort_values(ascending=False).head(10)

District Area
Blaeu Gwent          1.693694
West Somerset        1.613990
Derwentside          1.606186
Chester-le-Street    1.603306
Durham               1.580431
Blaenau Gwent        1.579310
Fylde                1.569350
Hambleton            1.568523
Merthyr Tydfil       1.568182
West Lancashire      1.566341
Name: Number_of_Casualties, dtype: float64

<h2>Insight: Blaeu Gwent has the highest average casualties per accident with 1.69, meaning accidents in this district tend to involve more people compared to other areas.</h2>

<h2>10. Correlation between casualties and latitude</h2>

In [52]:
df['Number_of_Casualties'].corr(df['Latitude'])

np.float64(0.032200686625906395)

<h2>Insight: This correlation shows whether accident severity (in terms of casualties) has any geographic pattern north or south. A weak or near-zero value means location has little effect on casualties.</h2>

<h2>11. Correlation between casualties and longitude</h2>

In [53]:
df['Number_of_Casualties'].corr(df['Longitude'])


np.float64(-0.04040564578845451)

<h2>Insight: This reveals if there is an east–west pattern in casualty numbers. A near-zero correlation suggests location doesn’t significantly influence accident casualties.</h2>

<h2>12. Correlation between vehicles and latitude)</h2>

In [56]:
df['Number_of_Vehicles'].corr(df['Latitude'])


np.float64(-0.04002766693447657)

<h2>Insight: This indicates whether multi-vehicle accidents are more common in northern vs southern areas. A low correlation would mean geography isn’t a strong factor.</h2>

<h2>13. Which vehicle type is most involved in fatal accidents?</h2>

In [30]:
df[df['Accident_Severity'] == 'Fatal']['Vehicle_Type'].value_counts().head(5)


Vehicle_Type
Car                                     6577
Van / Goods 3.5 tonnes mgw or under      467
Motorcycle over 500cc                    339
Bus or coach (17 or more pass seats)     325
Goods 7.5 tonnes mgw and over            216
Name: count, dtype: int64

<h2>Insight: The vehicle type most frequently involved in fatal accidents stands out as the riskiest on the road, suggesting it may need stricter safety measures, driver awareness programs, or targeted road safety policies.</h2>

<h2>14. What is the maximum casualties in a single accident?</h2>

In [34]:
max_casualty_signle_accident = df['Number_of_Casualties'].max()
print(max_casualty_signle_accident)

68


<h2>Insight: The maximum casualties recorded in a single accident show the most severe incident in this dataset, emphasizing the extreme impact such rare but devastating events can have on road safety.</h2>

<h2>15. Which accident severity has the highest average casualties?</h2>

In [36]:
df.groupby('Accident_Severity')['Number_of_Casualties'].mean().sort_values(ascending=False)


Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2>Insight: The accident severity category with the highest average casualties shows which type of crashes typically involve more people underscoring the deadliest type of incidents in terms of impact per accident.</h2>

<h2>16. Which road type records the most accidents?</h2>

In [37]:
df['Road_Type'].value_counts().head(5)


Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h2>Insight: The road type with the most accidents highlights where crashes most frequently occur, pointing to infrastructure or traffic conditions that may need closer attention for road safety improvements.</h2>

<h2>17. Do more accidents happen in urban or rural areas?</h2>

In [50]:
df['Urban_or_Rural_Area'].value_counts()



Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insight: Comparing accident counts between urban and rural areas reveals whether densely populated city roads or less populated rural roads see more crashes giving important context for where road safety measures should be prioritized.</h2>

<h2>18. Which area type (Urban/Rural) has higher average casualties?</h2>

In [45]:
df.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean().sort_values(ascending=False)


Urban_or_Rural_Area
Rural          1.479204
Urban          1.287807
Unallocated    1.181818
Name: Number_of_Casualties, dtype: float64

<h2>Insight: The area type (Urban or Rural) with the higher average casualties shows where accidents tend to be more severe on a per-incident basis, suggesting that while urban areas may have more accidents overall, rural crashes often involve more people per accident.</h2>

<h2>19. Correlation between casualties and vehicles</h2>

In [48]:
corr_casualties_vehicle = df['Number_of_Casualties'].corr(df['Number_of_Vehicles'])
corr_casualties_vehicle


np.float64(0.2288888612692756)

<h2>Insight: The correlation value here shows the strength and direction of the relationship between the number of vehicles and the number of casualties. A strong positive correlation would mean that multi-vehicle accidents tend to produce more casualties, while a weak correlation would suggest that other factors like severity or road type play a bigger role.</h2>

<h2>20. Correlation between vehicles and longitude</h2>

In [57]:
df['Number_of_Vehicles'].corr(df['Longitude'])


np.float64(0.014724875829090085)

<h2>Insight: This checks if the number of vehicles involved in accidents has any east–west trend. A weak correlation shows distribution is fairly even across regions.</h2>

<h2>21. Correlation between casualties and accident severity</h2>

In [62]:
df.groupby('Accident_Severity')['Number_of_Casualties'].mean().sort_values(ascending=False)


Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2>Insight: The grouped means directly show which severity level has higher average casualties (easy to interpret and avoids numeric encoding entirely).</h2>

<h2>22. Which district has the most fatal accidents?</h2>

In [49]:
df[df['Accident_Severity'] == 'Fatal']['District Area'].value_counts().head(5)


District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
Name: count, dtype: int64

<h2>Insight: The district with the most fatal accidents represents the area where deadly crashes are most concentrated, signaling a high-priority zone for road safety interventions and policy focus.</h2>